### PR TITLE
Tr/making the tests great

### PR DESCRIPTION
**Pull request by mrT23**
The real deal - implementing https://github.com/Codium-ai/cover-agent/issues/39 and https://github.com/Codium-ai/cover-agent/issues/40

image

### **PR Type**
Enhancement, Tests

___

### **Description**
- Enhanced the `generate_tests` and `validate_test` methods in `UnitTestGenerator.py` to handle JSON serialization, new imports, and specific line insertions.
- Updated `PromptBuilder.py` to include numbered test file content in the prompt.
- Modified the main function in `main.py` to handle the new dictionary structure returned by `generate_tests`.
- Updated the test generation prompt guidelines and model in `test_generation_prompt.toml`.
- Added new JavaScript classes (`UI` and `Github`) and HTML structure for the GitHub User Finder application.
- Added initial tests for the `UI` class using Vitest.
- Added package configuration, main application logic, and Vitest configuration for the js-vanilla-example project.
- Added a README file for the js-vanilla-example project.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><details><summary>9 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>UnitTestGenerator.py</strong><dd><code>Enhance test generation and validation logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cover_agent/UnitTestGenerator.py

<li>Modified the <code>generate_tests</code> method to handle JSON serialization of <br>failed test runs.<br> <li> Updated the <code>validate_test</code> method to handle new imports and insert <br>tests at specific lines.<br> <li> Added a new helper function <code>extract_error_message</code> for extracting error <br>messages.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-19760582d9ede3a799fdbb541ad357b4822682e837bca8365196fba50daf57e3">+132/-141</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>PromptBuilder.py</strong><dd><code>Include numbered test file content in prompt</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cover_agent/PromptBuilder.py

<li>Added numbering to the test file content.<br> <li> Included <code>test_file_numbered</code> in the prompt dictionary.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-1602223755506d62450bdc4cea9140b0a11ccf93f7cd00151970cd1d649786ad">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>main.py</strong><dd><code>Update main function for new test generation structure</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cover_agent/main.py

<li>Updated the main function to handle the new dictionary structure <br>returned by <code>generate_tests</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-610f77abc2025129e317ec1f98b3bb375b790b2197ed3f635331cca0a06fd999">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>test_generation_prompt.toml</strong><dd><code>Update test generation prompt guidelines and model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cover_agent/settings/test_generation_prompt.toml

<li>Updated guidelines to ensure new tests are part of the existing test <br>suite.<br> <li> Added fields for <code>relevant_line_to_insert_after</code> and <code>needed_indent</code> in <br>the <code>NewTests</code> model.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-760e793bfd5f739e4369c90b8347a1af81c1a2b57f9b7ce2487379fc18b0b926">+12/-6</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ui.js</strong><dd><code>Add UI class for handling user interactions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/js_vanilla/ui.js

- Added a new `UI` class to handle user interface interactions.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-1eddf3217173050bef30de57096e6a58d7568f9057362d9bbf40f1b2eca789cc">+82/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.html</strong><dd><code>Add HTML structure for GitHub User Finder</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/js_vanilla/index.html

- Added HTML structure for the GitHub User Finder application.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-631db1bb94fda5e4d587c8d768414f0959247ccd52a24fb77bf422fa58643ecd">+33/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>github.js</strong><dd><code>Add Github class for API interactions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/js_vanilla/github.js

- Added a new `Github` class to handle GitHub API interactions.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-8cb1647b07ef850521b03676a0dd6b13329c990b82923da3c83a1bb0c193c074">+25/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>package.json</strong><dd><code>Add package configuration for js-vanilla-example</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/js_vanilla/package.json

<li>Added package configuration for the js-vanilla-example project.<br> <li> Included dependencies and scripts for testing.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-748b28f32c180799c06838e4e185b47da5745489b28913ce458522419c3b2836">+20/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>app.js</strong><dd><code>Add main application logic for GitHub User Finder</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/js_vanilla/app.js

<li>Added main application logic for handling user input and displaying <br>GitHub user data.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-764879dae18fb50e6da5ddadf62a2faeff55bc5d2052fdcc285f74741ab2bb4c">+22/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests
</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>ui.test.js</strong><dd><code>Add initial tests for UI class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/js_vanilla/ui.test.js

- Added initial tests for the `UI` class using Vitest.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-0749ad53207b941a7439a3601e9364d0f30c573761479a0114d2b5089b629ad9">+29/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>vitest.config.js</strong><dd><code>Add Vitest configuration for testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/js_vanilla/vitest.config.js

- Added configuration for Vitest testing framework.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-bc26906efbdc2565427a1976ee7b5356cf15e47a35f9850ccce1809bd655a635">+13/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Add README for js-vanilla-example project</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
templated_tests/js_vanilla/README.md

- Added README file for the js-vanilla-example project.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/50/files#diff-1769bb6293737e902535dc33111b5e76b72a1309d9aa02ec6b2184c1a244a22b">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by amahpour** [comment]
@mrT23 can we add more unit tests to bring coverage back up to 70%?

**Comment by EmbeddedDevops1** [comment]
Looks like CI is failing because code coverage dropped by 70%. How about using this branch to extend coverage? 😄

File: cover_agent/PromptBuilder.py [hunk 1]
```diff
@@ -125,6 +127,7 @@ def build_prompt(self) -> dict:
             "source_file_name": self.source_file_name,
             "test_file_name": self.test_file_name,
             "source_file_numbered": self.source_file_numbered,
+            "test_file_numbered": self.test_fule_numbered,
```
Comment (EmbeddedDevops1) [inline cover_agent/PromptBuilder.py:130]: Spelling error: `fule` but should be `file`. Need to refactor.

File: templated_tests/js_vanilla/github.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+class Github {
+  constructor() {
+    this.client_id = 'ab7e3c6ac10d3714249a';
+    this.client_secret = 'f315c3cc4bca8b4b922fc04af1b31b02cb1d143d';
```
Comment (EmbeddedDevops1) [inline templated_tests/js_vanilla/github.js:4]: Are these real or fake values?

File: cover_agent/UnitTestGenerator.py [hunk 3]
```diff
@@ -201,14 +201,18 @@ def build_prompt(self):
             failed_test_runs_value = ""
             try:
                 for failed_test in self.failed_test_runs:
-                    code = failed_test['code'].strip()
+                    failed_test_dict = failed_test['code']
```
Comment (EmbeddedDevops1) [inline cover_agent/UnitTestGenerator.py:204]: Suggested by PR Agent:
```suggestion
                    failed_test_dict = failed_test.get('code', {})
```

File: cover_agent/main.py [hunk 4]
```diff
@@ -160,14 +160,14 @@ def main():
...
-            generated_tests = test_gen.generate_tests(max_tokens=4096)
+            generated_tests_dict = test_gen.generate_tests(max_tokens=4096)
 
             # Write test_gen.prompt to a debug markdown file
             write_prompt_to_file(GENERATED_PROMPT_NAME, test_gen.prompt)
 
             # Validate each test and append the results to the test results list
-            for generated_test in generated_tests:
-                test_result = test_gen.validate_test(generated_test)
+            for generated_test in generated_tests_dict['tests']:
```
Comment (EmbeddedDevops1) [inline cover_agent/main.py:169]: Suggestion from PR Agent
```suggestion
            for generated_test in generated_tests_dict.get('tests', []):
```

File: templated_tests/js_vanilla/github.js (hunk 2 above)
Comment (mrT23) [reply templated_tests/js_vanilla/github.js:4]: fake

File: cover_agent/PromptBuilder.py (hunk 1 above)
Comment (mrT23) [reply cover_agent/PromptBuilder.py]: fixed

File: cover_agent/UnitTestGenerator.py (hunk 3 above)
Comment (mrT23) [reply cover_agent/UnitTestGenerator.py]: done

File: cover_agent/main.py (hunk 4 above)
Comment (mrT23) [reply cover_agent/main.py]: done

**Pull request merged**